manager/perf.go

```diff
@@ -0,0 +1,156 @@
+package manager
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/DataDog/ebpf"
+	"github.com/DataDog/ebpf/perf"
+)
+
+// PerfMapOptions - Perf map specific options
+type PerfMapOptions struct {
+	// PerfRingBufferSize - Size in bytes of the perf ring buffer. Defaults to the manager value if not set.
+	PerfRingBufferSize int
+
+	// Watermark - The reader will start processing samples once their sizes in the perf ring buffer
+	// exceed this value. Must be smaller than PerfRingBufferSize. Defaults to the manager value if not set.
+	Watermark int
+
+	// DataHandler - Callback function called when a new sample was retrieved from the perf
+	// ring buffer.
+	DataHandler func(CPU int, data []byte, perfMap *PerfMap, manager *Manager)
+
+	// LostHandler - Callback function called when one or more events where dropped by the kernel
+	// because the perf ring buffer was full.
+	LostHandler func(CPU int, count uint64, perfMap *PerfMap, manager *Manager)
+}
+
+// PerfMap - Perf ring buffer reader wrapper
+type PerfMap struct {
+	manager     *Manager
+	perfReader  *perf.Reader
+	running     bool
+	runningLock *sync.RWMutex
+
+	// Map - A PerfMap has the same features as a normal Map
+	Map
+	PerfMapOptions
+}
+
+// loadNewPerfMap - Creates a new perf map instance, loads it and setup the perf ring buffer reader
+func loadNewPerfMap(spec ebpf.MapSpec, options MapOptions, perfOptions PerfMapOptions) (*PerfMap, error) {
+	// Create underlying map
+	innerMap, err := loadNewMap(spec, options)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the new map
+	perfMap := PerfMap{
+		Map: *innerMap,
+		PerfMapOptions: perfOptions,
+	}
+	return &perfMap, nil
+}
+
+// Init - Initialize a map
+func (m *PerfMap) Init(manager *Manager) error {
+	m.runningLock = &sync.RWMutex{}
+	m.manager = manager
+
+	// Initialize the underlying map structure
+	if err := m.Map.Init(manager); err != nil {
+		return err
+	}
+
+	// Set default values if not already set
+	if m.PerfRingBufferSize == 0 {
+		m.PerfRingBufferSize = manager.options.DefaultPerfRingBufferSize
+	}
+	if m.Watermark == 0 {
+		m.Watermark = manager.options.DefaultWatermark
+	}
+
+	if m.DataHandler == nil {
```

Comment:
you could move this at the very beginning of the function

Reply:
👍